woocommerce/woocommerce-ios
# Products: Support products and variations with decimal stock quantities

Related issue: #3494 
Related PR: https://github.com/woocommerce/woocommerce-ios/pull/3707

## Description

The WooCommerce Core API only supports stock quantities with `Int` or `nil` values. However, certain third-party extensions can enable decimal (non-integer) stock quantities. When products have decimal stock quantities, the list of products on the Products tab doesn't load (current behavior in `develop`).

This PR adds support for decimal stock quantities to be displayed in the app, so the product list will load.

Note: This does not add full support for editing products with decimal stock quantities (which requires an API change). For now, we will restrict products with decimal stock quantities to be read-only in the app; that change is in https://github.com/woocommerce/woocommerce-ios/pull/3707 along with a release note about the read-only support. Internal ref: p91TBi-4iF-p2#comment-3111

## Changes

* Adds a new Core Data model 46, which changes `ProductVariation.stockQuantity` from `Int64` to `Decimal`. (`Product.stockQuantity` is already stored as a `String` so it doesn't require any Core Data changes.)
* Adds a mapping model for Model 45 to Model 46.
* Adds a unit test for the Model 45 to Model 46 migration, confirming that the attribute type is changed as expected.
* Updates `stockQuantity` throughout the app from `Int64` to `Decimal`.

Before|After - Products|After - Product Variations
-|-|-
![Simulator Screen Shot - iPhone 12 - 2021-02-19 at 16 26 33](https://user-images.githubusercontent.com/8658164/108539812-618c0200-72d8-11eb-814a-fc1caa8779b6.png)|![Simulator Screen Shot - iPhone 12 - 2021-02-19 at 16 25 16](https://user-images.githubusercontent.com/8658164/108539820-63ee5c00-72d8-11eb-9ea3-a0c39cb541ef.png)|![Simulator Screen Shot - iPhone 12 - 2021-02-19 at 16 46 12](https://user-images.githubusercontent.com/8658164/108540838-96e51f80-72d9-11eb-987a-3c4a8e5a3287.png)



## Testing

1. Switch to `develop`. Clean, build, and launch the app.
2. Log in and open the Products tab to fetch products. Open a variable product and open Variations to fetch the variations for that product.
3. On your store, enable decimal stock quantities by installing a third-party extension on your store like [WPC Product Quantity](https://wordpress.org/plugins/wpc-product-quantity/).
4. Update the inventory for at least one product and one product variation to have a decimal stock quantity.
5. Open the app, refresh the Products tab, and notice the products list doesn't load ("Unable to refresh list").
6. Kill the app.
7. Switch to this branch. Clean, rebuild, and launch the app.
8. Open the Products tab and confirm the products list now loads with no errors, showing the correct stock quantity for your product with a decimal stock quantity.
9. Open a variable product with at least one variation with a decimal stock quantity.
10. Open "Variations" and confirm the variations list loads, showing the correct stock quantity for the variation with a decimal stock quantity.

## Submitter Checklist

Update release notes:
